### PR TITLE
Algorithm - Add top-1 load balancing loss for MoE routing

### PR DESCRIPTION
**Pull request by qghuxmu**
Adds a new load balancing loss optimized for top-1 MoE routing. This loss mitigates the issue that even with imbalanced routing frequencies, the value of existing load balance loss could remain low.

### CODE REVIEW

File: megatron/training/training.py [hunk 1]
```diff
@@ -1538,6 +1538,11 @@ def training_log(loss_dict, total_loss_dict, learning_rate, decoupled_learning_r
                 track_names.append("global_batch_tokens_per_expert")
         if args.moe_z_loss_coeff is not None:
             track_names.append("z_loss")
+        if args.moe_onehot_lbl_coeff > 0:
+            track_names.append("onehot_load_balancing_loss")
```
Comment (yzygitzh) [inline megatron/training/training.py]: Merge onehot lbl to lbl loss category, and remove the following three

File: megatron/training/training.py [hunk 2]
```diff
@@ -1534,7 +1534,7 @@ def training_log(loss_dict, total_loss_dict, learning_rate, decoupled_learning_r
                 track_names.append("aux_tokens_per_expert")
             elif args.moe_router_load_balancing_type == "seq_aux_loss":
                 track_names.append("seq_aux_tokens_per_expert")
-            elif args.moe_router_load_balancing_type == "global_batch_loss":
+            elif args.moe_router_load_balancing_type in ["global_batch_loss", "top1_loss"]:
```
Comment (yzygitzh) [inline megatron/training/training.py]: Please create a new `elif` and add `track_names.append("topk1_tokens_per_expert")`

File: megatron/training/arguments.py [hunk 3]
```diff
@@ -2553,9 +2553,9 @@ def _add_moe_args(parser):
...
-                       help='Determines the load balancing strategy for the router. "aux_loss" corresponds to the load balancing loss used in GShard and SwitchTransformer; "seq_aux_loss" corresponds to the load balancing loss used in DeepSeekV2, which computes the loss for each individual sample; "sinkhorn" corresponds to the balancing algorithm used in S-BASE; "global_batch_loss" corresponds to the global-batch load balancing loss (see https://arxiv.org/abs/2501.11873 for details), and "none" implies no load balancing. The default is "aux_loss".')
+                       help='Determines the load balancing strategy for the router. "aux_loss" corresponds to the load balancing loss used in GShard and SwitchTransformer; "seq_aux_loss" corresponds to the load balancing loss used in DeepSeekV2, which computes the loss for each individual sample; "sinkhorn" corresponds to the balancing algorithm used in S-BASE; "global_batch_loss" corresponds to the global-batch load balancing loss (see https://arxiv.org/abs/2501.11873 for details), "top1_loss" corresponds to the top-1 load balancing loss, and "none" implies no load balancing. The default is "aux_loss".')
```
Comment (yzygitzh) [inline megatron/training/arguments.py]: Please change the same comment block in for this argument in `transformer_config.py` as well

File: megatron/training/arguments.py [hunk 4]
```diff
@@ -2601,6 +2601,8 @@ def _add_moe_args(parser):
                        help='Scaling coefficient for the z-loss: a starting value of 1e-3 is recommended.')
     group.add_argument('--moe-input-jitter-eps', type=float, default=None,
                        help='Add noise to the input tensor by applying jitter with a specified epsilon value.')
+    group.add_argument('--moe-top1-loss-temperature', type=float, default=1.0,
```
Comment (yzygitzh) [inline megatron/training/arguments.py]: Please move this after `--moe-router-load-balancing-type`

File: megatron/training/arguments.py [hunk 5]
```diff
@@ -2601,6 +2601,8 @@ def _add_moe_args(parser):
...
+                       help='Temperature for the top-1 load balancing loss: a starting value of 0.1 is recommended.')
```
Comment (yzygitzh) [inline megatron/training/arguments.py]: Please align the help comment with the argument in `transformer_config.py`

File: megatron/core/transformer/transformer_config.py [hunk 6]
```diff
@@ -431,6 +431,9 @@ class TransformerConfig(ModelParallelConfig):
     moe_input_jitter_eps: Optional[float] = None
     """Add noise to the input tensor by applying jitter with a specified epsilon value."""
 
+    moe_top1_loss_temperature: float = 1.0
```
Comment (yzygitzh) [inline megatron/core/transformer/transformer_config.py]: Move this after `moe_router_load_balancing_type`

File: megatron/core/transformer/moe/router.py [hunk 7]
```diff
@@ -370,6 +371,56 @@ def global_batch_loss_load_balancing(self, logits: torch.Tensor):
...
+                routing_map=routing_map,
+                topk=self.topk,
+            )
+            probs = self.apply_load_balancing_loss(
+                activation=probs, load_balancing_loss_func=aux_loss_func
+            )
+            
+            if self.config.moe_tokens_logging:
+                save_to_tokens_per_expert_tracker(
+                    "global_batch_tokens_per_expert",
```
Comment (yzygitzh) [inline megatron/core/transformer/moe/router.py]: Change name to `top1_batch_tokens_per_expert`

File: megatron/core/transformer/moe/moe_utils.py [hunk 8]
```diff
@@ -167,6 +167,56 @@ def global_batch_load_balancing_loss_func(
...
 
+def top1_load_balancing_loss_func(
+    probs: torch.Tensor,
+    routing_map: torch.Tensor,
+    topk: int,
+    moe_aux_loss_coeff: float,
+    sequence_partition_group=None,
+):
+    """
+    Calculate the top-1 auxiliary loss.
```
Comment (yzygitzh) [inline megatron/core/transformer/moe/moe_utils.py]: Please add brief description about why this is better for top1

File: megatron/core/transformer/moe/moe_utils.py [hunk 9]
```diff
@@ -167,6 +167,56 @@ def global_batch_load_balancing_loss_func(
...
+    Returns:
+        torch.Tensor: The sequence auxiliary loss for load balancing.
+    """
+    num_experts = probs.shape[-1]
+
+    differentiable_f = probs.mean(dim=0)
+
+    topk_scores_per_token = probs[routing_map].view(-1, topk)
+    sum_topk_scores_per_token = topk_scores_per_token.sum(dim=-1)
+    mean_topk_scores = sum_topk_scores_per_token.mean()
```
Comment (yzygitzh) [inline megatron/core/transformer/moe/moe_utils.py]: Shall we specify `.mean(dim=-1)` here to make it clear

Comment (qghuxmu) [reply megatron/core/transformer/moe/moe_utils.py]: sure

**Comment by yzygitzh** [comment]
Also please help add a test case in tests/unit_tests/transformer/moe/test_aux_loss.py @qghuxmu

**Review by yzygitzh — DISMISSED** [review]

**Review by yzygitzh — APPROVED** [review]

File: megatron/core/transformer/transformer_config.py [hunk 10]
```diff
@@ -344,9 +344,11 @@ class TransformerConfig(ModelParallelConfig):
     used in GShard and SwitchTransformer; "seq_aux_loss" corresponds to the loss used in DeepSeekV2,
     which computes the loss for each individual sample; "sinkhorn" corresponds to the balancing
     algorithm used in S-BASE; "global_batch_loss" corresponds to the global batch load balancing loss
-    (see https://arxiv.org/abs/2501.11873 for details), and "none" implies no load balancing. The
-    default is "aux_loss"."""
+    (see https://arxiv.org/abs/2501.11873 for details); "top1_loss" corresponds to the top-1 load balancing loss,
+    and "none" implies no load balancing. The default is "aux_loss"."""
 
+    moe_top1_loss_temperature: float = 1.0
```
Comment (yukirora) [inline megatron/core/transformer/transformer_config.py]: please add the assert to filter out invalid value

**Review by yukirora — APPROVED** [review]
